File: apps/clinical-engine/src/ehas2_clinical_engine/rule4/eligibility/formula_bp_gate.py
from __future__ import annotations

from typing import Literal, TypedDict


class FormulaBpReading(TypedDict, total=False):
    formula_slot_id: str
    formula_target_id: str
    systolic_mm_hg: float
    diastolic_mm_hg: float
    unit: Literal["mmHg"]
    verification_status: str
    reading_role: str
    evidence_item_id: str | None
    organ_target_binding_status: str
    cardiac_target_class: str | None


class FormulaBpGateResult(TypedDict):
    outcome: Literal["PASS", "FAIL", "MISSING_INPUT", "CONTRADICTORY"]
    reason_codes: list[str]


def _mm_hg_numeric(value: object) -> float | None:
    if isinstance(value, bool):
        return None
    if isinstance(value, int | float):
        return float(value)
    return None


def evaluate_formula_bp_stage_gate(
    reading: FormulaBpReading | None,
    required_stage: Literal[1, 2],
) -> FormulaBpGateResult:
    if not reading:
        return {"outcome": "MISSING_INPUT", "reason_codes": ["FORMULA_BP_READING_MISSING"]}
    if reading.get("unit") != "mmHg":
        return {"outcome": "MISSING_INPUT", "reason_codes": ["FORMULA_BP_NOT_VERIFIED"]}
    if reading.get("verification_status") != "VERIFIED":
        return {"outcome": "MISSING_INPUT", "reason_codes": ["FORMULA_BP_NOT_VERIFIED"]}
    if reading.get("organ_target_binding_status") == "MISMATCH":
        return {"outcome": "CONTRADICTORY", "reason_codes": ["FORMULA_BP_TARGET_MISMATCH"]}
    if reading.get("organ_target_binding_status") == "MISSING":
        return {"outcome": "MISSING_INPUT", "reason_codes": ["FORMULA_BP_BINDING_MISSING"]}
    sys_val = _mm_hg_numeric(reading.get("systolic_mm_hg"))
    dia_val = _mm_hg_numeric(reading.get("diastolic_mm_hg"))
    if sys_val is None or dia_val is None:
        return {"outcome": "MISSING_INPUT", "reason_codes": ["FORMULA_BP_READING_MISSING"]}
    sys = sys_val
    dia = dia_val
    if required_stage == 1:
        stage1 = (140 <= sys <= 159) or (90 <= dia <= 99)
        if stage1:
            return {"outcome": "PASS", "reason_codes": []}
        return {"outcome": "FAIL", "reason_codes": ["FORMULA_BP_STAGE1_NOT_MET"]}
    stage2 = (160 <= sys <= 179) or (100 <= dia <= 109)
    if stage2:
        return {"outcome": "PASS", "reason_codes": []}
    return {"outcome": "FAIL", "reason_codes": ["FORMULA_BP_STAGE2_NOT_MET"]}

File: packages/clinical-contracts/src/rule4/eligibility/formulaBpGate.ts
import type { Rule4FormulaBpReading } from './types.js';

/** Q06C Stage 1: sys 140–159 or dia 90–99. Stage 2: sys 160–179 or dia 100–109. Crisis handled in Phase 2. */
export function evaluateFormulaBpStageGate(
  reading: Rule4FormulaBpReading | null | undefined,
  requiredStage: 1 | 2,
): {
  outcome: 'PASS' | 'FAIL' | 'MISSING_INPUT' | 'CONTRADICTORY';
  reasonCodes: string[];
} {
  if (!reading) {
    return { outcome: 'MISSING_INPUT', reasonCodes: ['FORMULA_BP_READING_MISSING'] };
  }
  if (reading.unit !== 'mmHg') {
    return { outcome: 'MISSING_INPUT', reasonCodes: ['FORMULA_BP_NOT_VERIFIED'] };
  }
  if (reading.verificationStatus !== 'VERIFIED') {
    return { outcome: 'MISSING_INPUT', reasonCodes: ['FORMULA_BP_NOT_VERIFIED'] };
  }
  if (reading.organTargetBindingStatus === 'MISMATCH') {
    return { outcome: 'CONTRADICTORY', reasonCodes: ['FORMULA_BP_TARGET_MISMATCH'] };
  }
  if (reading.organTargetBindingStatus === 'MISSING') {
    return { outcome: 'MISSING_INPUT', reasonCodes: ['FORMULA_BP_BINDING_MISSING'] };
  }
  const sys = reading.systolicMmHg;
  const dia = reading.diastolicMmHg;
  if (requiredStage === 1) {
    const stage1 = (sys >= 140 && sys <= 159) || (dia >= 90 && dia <= 99);
    return stage1
      ? { outcome: 'PASS', reasonCodes: [] }
      : { outcome: 'FAIL', reasonCodes: ['FORMULA_BP_STAGE1_NOT_MET'] };
  }
  const stage2 = (sys >= 160 && sys <= 179) || (dia >= 100 && dia <= 109);
  return stage2
    ? { outcome: 'PASS', reasonCodes: [] }
    : { outcome: 'FAIL', reasonCodes: ['FORMULA_BP_STAGE2_NOT_MET'] };
}
